tracing/AsyncWrap/conceptual_overview.md

```diff
@@ -0,0 +1,73 @@
+
+When a task (callback) is registered, a context ID is assigned and provided as a parameter to "init" hooks. Before the task (callback) is invoked the same context ID is provided to "before" hooks. "init" hooks execute immediately in the registering context and therefore can save off aspects of it; "before" hooks execute in the task invocation context and therefore can restore saved aspects of the registering context prior to that task's execution.
+
+"after" hooks are called immediately after execution of the task and receive the context ID to allow context cleanup. "destroy" hooks are called on destruction of the context to allow resource cleanup.
+
+As implied by the previous paragraph, async\_hooks itself only manages a context ID, not actual context. It is up to a hook author to associate data with this ID in an init hook, for example in a persistent map of context IDs to context objects.
+
+## Terms
+
+**async\_id** is a unique ID assigned to every context. An incrementing integer; 1 represents the root context.
+
+**trigger\_id** is the async\_id of the parent context of the context identified by async\_id.
+
+**registering context** is the context where a task is scheduled, for example where `nextTick` is called.
+
+**invocation context** is the context where a task is invoked.
+
+## Implementations
+
+The above concepts apply to all implementations of async\_hooks; but the implementation is different for different task scheduling mechanisms.
+
+The implementation of async\_hooks integrated in Node.js core focuses mostly on tasks and callbacks associated with UV resources. It also includes an implementation for `process.nextTick` and a JS API for use by library authors. An implementation for use with Promises which utilizes PromiseHook is also in progress.
+
+### UV Resources
+
+For UV resources (handles and requests), async\_hooks provides a mechanism to track a context ID across the lifetime of the UV resource. When the UV resource is constructed a context ID is created and "init" hooks called; when callbacks associated with the resource are executed this same context ID is passed to "before" hooks. Construction and destruction of the UV resource are taken to be adequate proxies for construction and destruction of a logical context.
+
+### JS Libraries
+
+An early form of an "Embedder API" is available in the `async\_hooks` JS module which allows a library author to create a new context (i.e. async\_id) independently of a UV resource. It is up to the library author to call `emitInit` to create the new context at the time of task registration, and to wrap the callback with `emitBefore` and `emitAfter` for execution at the time invocation. The library author can also call emitDestroy when the original context is no longer needed.
```

Comment:
The JS API is not just a predecessor (early from) of the C++ API, it is useful in cases where asynchronous flow doesn't reflect the causal chain of events, an example is a queue.